compiler/cpp/src/lexer/Lexer.cpp

```diff
@@ -0,0 +1,279 @@
+    currentContext = PAYLOAD;
+    return constructToken(TOK_BRACE_OPEN);
+}
+
+Token Lexer::consumeBraceClose() {
+    reader.expect('}');
+    if (!isLookAheadCommentBlockCloseIdentifier()) currentContext = ARBITRARY;
+    return constructToken(TOK_BRACE_CLOSE);
+}
+
+Token Lexer::consumeIndex() {
+    reader.expect('[');
+    Token indexToken = consumeNumber();
+    reader.expect(']');
+    return constructToken(TOK_INDEX, indexToken.getValue());
+}
+
+Token Lexer::consumeString() {
+    reader.expect('"');
+    std::stringstream stringStr;
+    while(reader.getLookahead() != '"' && reader.getLookahead() != EOF) {
+        if (reader.getLookahead() == '\\') reader.advance();
+        stringStr << (char) reader.getLookahead();
+        reader.advance();
+    }
+    reader.expect('"');
+    return constructToken(TOK_STRING, stringStr.str());
+}
+
+Token Lexer::consumeIdentifierOrKeyword() {
+    std::stringstream identifierStr;
+    identifierStr << (char) reader.getLookahead();
+    reader.advance();
+    while (isalnum(reader.getLookahead())) { // [a-zA-Z0-9]*
+        identifierStr << (char) reader.getLookahead();
+        reader.advance();
+    }
+
+    // Is keyword?
+    std::string identifierName = identifierStr.str();
+    if (identifierName == "if") return constructToken(TOK_IF);
+    if (identifierName == "has") return constructToken(TOK_HAS);
+    if (identifierName == "not") return constructToken(TOK_NOT);
+    if (identifierName == "true") return constructToken(TOK_TRUE);
+    if (identifierName == "false") return constructToken(TOK_FALSE);
+
+    return constructToken(TOK_IDENTIFIER, identifierName);
+}
+
+Token Lexer::consumeNumber() {
+    std::stringstream numStr;
+    do {
+        numStr << (char) reader.getLookahead();
+        reader.advance();
+    } while (isdigit(reader.getLookahead()));
+    return constructToken(TOK_NUMBER, numStr.str());
+}
+
+Token Lexer::consumeCommentLineIdentifier() {
+    reader.expectMultiple(commentLineIdentifier);
+    currentContext = SECTION;
+    return constructToken(TOK_COM_LINE_IDEN);
+}
+
+Token Lexer::consumeCommentBlockOpenIdentifier() {
+    reader.expectMultiple(commentBlockOpenIdentifier);
+    currentContext = SECTION;
+    return constructToken(TOK_COM_BLOCK_IDEN_OPEN);
+}
+
+Token Lexer::consumeCommentBlockCloseIdentifier() {
+    reader.expectMultiple(commentBlockCloseIdentifier);
+    currentContext = ARBITRARY;
+    return constructToken(TOK_COM_BLOCK_IDEN_CLOSE);
+}
+
+Token Lexer::consumeCommentPayloadIdentifier() {
+    reader.expectMultiple(commentPayloadIdentifier);
+    return constructToken(TOK_COM_IDEN_PAYLOAD);
+}
+
+// -------------------------------------------------- Util functions ---------------------------------------------------
+
+bool Lexer::isEOF() {
+    return reader.getLookahead() == EOF;
+}
+
+bool Lexer::isLookAheadCommentLineIdentifier() {
+    return !commentLineIdentifier.empty() && reader.getMaxLookahead().find(commentLineIdentifier) == 0;
```

Comment:
I suggest to rename `getMaxLookahead()` since - given your method name -  I would expect to get an int back describing the max look ahead. Maybe something along the lines of `getLookAheadMultiple()` or `getNextChars()` etc. ?

Reply:
Good idea, will apply that.